DirectProgramming/DPC++FPGA/ReferenceDesigns/qri/src/qri_demo.cpp

```diff
@@ -0,0 +1,580 @@
+    %
+    % OUTPUT:
+    %    A        a matrix which is strictly diagonally domimant by rows
+    %    B        B = D\A, where D is the diagonal of A
+    %
+    % The main purpose of this function is to construct test matrices
+    % for, say, Gaussian elimination with no pivoting.
+    %
+    % The matrix A is not necessarily well-conditioned, but
+    % the infinity norm condition number of the matrix B is bounded by
+    %
+    %                  (1+epsilon)/(1 - epsilon)
+
+    % PROGRAMMING by Carl Christian Kjelgaard Mikkelsen (spock@cs.umu.se)
+    %   2021-10-21  Initial programming and testing.
+
+    % Generate a random matrix
+    R=rand(m,m)-rand(m,m);
+
+    % Eliminate the diagonal
+    D=diag(diag(R)); R=R-D;
+
+    % Measure the weight of the off diagonal entries
+    w=abs(R)*ones(m,1);
+
+    % Construct the new diagonal elements
+    d=w/epsilon;
+
+    % Construct the matrix which is diagonally dominant
+    A=R+diag(d);
+
+    % Do the diagonal scaling
+    B=diag(diag(A))\A;
+*/
+template <int size, typename T>
+void GenerateMatrixWithCondititionNumber(float epsilon,
+                                         std::vector<T> &output) {
+  // Random min and max values for the random floating-point value generation
+  constexpr float kRandomMin = 0;
+  constexpr float kRandomMax = 1;
+
+  // Generate a random matrix R with diagonal elements set to 0
+  // and measure the weights of the off diagonal entries
+  std::vector<T> r, weights;
+  r.resize(size * size);
+  weights.resize(size);
+  for (int row = 0; row < size; row++) {
+    weights[row] = {0};
+    for (int col = 0; col < size; col++) {
+      if (col != row) {
+        float random1 = RandomValueInInterval(kRandomMin, kRandomMax);
+        T elem;
+#if COMPLEX == 1
+        float random1I = RandomValueInInterval(kRandomMin, kRandomMax);
+        elem = {random1, random1I};
+        r[row * size + col] = elem;
+#else
+        elem = random1;
+        r[row * size + col] = elem;
+#endif
+        weights[row] += elem;
+      }
+    }
+
+    // Construct the new diagonal element
+    weights[row] /= epsilon;
+    r[row * size + row] = weights[row];
+  }
+
+  // Perform the diagonal scaling by solving:
+  // diag(diag(A))*output = A
+  for (int row = 0; row < size; row++) {
+    for (int col = 0; col < size; col++) {
+      output[row * size + col] = r[row * size + col] / r[row * size + row];
+    }
+  }
+}
+
+int main(int argc, char *argv[]) {
+  constexpr size_t kRandomSeed = 1138;
+  constexpr size_t kRows = ROWS_COMPONENT;
+  constexpr size_t kColumns = COLS_COMPONENT;
+  constexpr size_t kAMatrixSize = kRows * kColumns;
+  constexpr size_t kInverseMatrixSize = kRows * kColumns;
+  constexpr bool kComplex = COMPLEX != 0;
+  constexpr size_t kMatricesToInvert = 8;
+
+  // Get the number of times we want to repeat the inversion
+  // from the command line.
+#if defined(FPGA_EMULATOR)
+  int repetitions = argc > 1 ? atoi(argv[1]) : 16;
```

Comment:
What happens if the user types in a bogus argument (-1) or FUBAR?

Reply:
This is what happens: 
```
>  ./qri.fpga_emu -1
Number of repetitions given is lower that 1.
The decomposition must occur at least 1 time.
Increase the number of repetitions (e.g. 16).
>  ./qri.fpga_emu FUBAR
Number of repetitions given is lower that 1.
The decomposition must occur at least 1 time.
Increase the number of repetitions (e.g. 16).

```
Admittedly it is not perfect, but it won't run unless the user provides a number >=1